core_backend/app/contents/models.py

```diff
@@ -0,0 +1,212 @@
+        updated_datetime_utc=datetime.utcnow(),
+    )
+
+    asession.add(content_db)
+
+    await asession.commit()
+    await asession.refresh(content_db)
+
+    return content_db
+
+
+async def update_content_in_db(
+    content_id: int,
+    content: ContentCreate,
+    asession: AsyncSession,
+) -> ContentDB:
+    """
+    Updates a content and vector in the database
+    """
+
+    content_embedding = await _get_content_embeddings(content)
+    content_db = ContentDB(
+        content_id=content_id,
+        content_embedding=content_embedding,
+        content_title=content.content_title,
+        content_text=content.content_text,
+        content_language=content.content_language,
+        content_metadata=content.content_metadata,
+        updated_datetime_utc=datetime.utcnow(),
+    )
+
+    content_db = await asession.merge(content_db)
+    await asession.commit()
+    return content_db
+
+
+async def delete_content_from_db(
+    content_id: int,
+    asession: AsyncSession,
+) -> None:
+    """
+    Deletes a content from the database
+    """
+    stmt = delete(ContentDB).where(ContentDB.content_id == content_id)
+    await asession.execute(stmt)
+    await asession.commit()
+
+
+async def get_content_from_db(
+    content_id: int,
+    asession: AsyncSession,
+) -> Optional[ContentDB]:
+    """
+    Retrieves a content from the database
+    """
+    stmt = select(ContentDB).where(ContentDB.content_id == content_id)
+    content_row = (await asession.execute(stmt)).first()
+    if content_row:
+        return content_row[0]
+    else:
+        return None
+
+
+async def get_list_of_content_from_db(
+    asession: AsyncSession, offset: int = 0, limit: Optional[int] = None
+) -> List[ContentDB]:
+    """
+    Retrieves all content from the database
+    """
+    stmt = select(ContentDB)
+    if offset > 0:
+        stmt = stmt.offset(offset)
+    if limit is not None:
+        stmt = stmt.limit(limit)
+    content_rows = (await asession.execute(stmt)).all()
+
+    return [c[0] for c in content_rows] if content_rows else []
+
+
+async def _get_content_embeddings(
+    content: ContentCreate | ContentUpdate,
+) -> List[float]:
+    """
+    Vectorizes the content
+    """
+    text_to_embed = content.content_title + "\n" + content.content_text
+    content_embedding = embedding(EMBEDDING_MODEL, text_to_embed).data[0]["embedding"]
+    return content_embedding
+
+
+async def get_similar_content(
+    question: str,
+    n_similar: int,
+    asession: AsyncSession,
+) -> Dict[int, tuple[Any, Any, Any]]:
+    """
+    Get the most similar points in the vector table
+    """
+    response = embedding(EMBEDDING_MODEL, question)
+    question_embedding = response.data[0]["embedding"]
+
+    return await get_search_results(
+        question_embedding,
+        n_similar,
+        asession,
+    )
+
+
+async def get_similar_content_async(
+    question: str, n_similar: int, asession: AsyncSession
```

Comment:
Made some changes for this function. We shouldn't use a  submodules' schema as input of outputs for another submodule's function. That could lead to circular imports. I modified it to take a `string` instead of `UserQueryBase` as  input and as output a `dict` rather than `UserQueryRespone`

Reply:
Why would there be circular imports? Schema file doesn't import other functionality right? But I do like the idea of not using schemas from other submodules where possible - less coupling.

Can we make the type `tuple[string, string, float]` or whatever it is instead of Any?

Reply:
It would be because  the contents module was using `UserQueryBase` from question_answer and question_answer would use `get_similar_content_async ` from contents.  That is why I made the change.